test/test_helper.rb

```diff
@@ -88,4 +95,34 @@ def assert_equivalent_xml(expected, actual)
     }
   end
 
+  ##
+  # Test with the @builder, and with @form_with_builder if Rails 5.1.0+
+  def assert_with_builder(expected, method, *args)
+    assert_equivalent_xml expected, @builder.send(method, *args)
+    if Gem::Version.new(::Rails::VERSION::STRING) >= Gem::Version.new('5.1.0')
+      assert_equivalent_xml remove_default_ids_for_rails_5_1(expected),
+                            @form_with_builder.send(method, *args)
+    end
+  end
+
+  def assert_with_builder_radio(expected, *args)
+    assert_equivalent_xml expected, @builder.radio_button(*args)
+    if Gem::Version.new(::Rails::VERSION::STRING) >= Gem::Version.new('5.1.0')
+      assert_equivalent_xml remove_default_ids_for_rails_5_1(expected),
+                            @form_with_builder.radio_button(*args)
+    end
+  end
+
+  ##
+  # Remove ids on labels if running on Rails 5.1.0
+  def remove_default_ids_for_rails_5_1(expected)
+    return expected unless Gem::Version.new('5.1.0') <= Gem::Version.new(::Rails::VERSION::STRING) && Gem::Version.new(::Rails::VERSION::STRING) < Gem::Version.new('5.2.0.beta2')
+    root = Nokogiri::XML(expected)
+
+    # There are more elegant ways to do this, I'm sure, but later for that.
+    root.traverse do |node|
+      node.delete("id") if node.has_attribute?("id")
+    end
+    root.to_s
+  end
```

Comment:
Trying to understand what these assertion helpers are for. Why `assert_equivalent_xml` helper is not enough?

Reply:
I have a dream that we can implement one or two helpers that cover the range of variations in output from the Rails helpers across Rails versions. Here's a summary of differences that I'm aware of:

* Prior to Rails 5, tabs were output in alphabetical order. After, they weren't
* Some helpers starting with Rails 5 produced more output, e.g. an empty hidden input on file tags (The genesis of `assert_equivalent_xml` is found in #273 and #278, which specifically addressed these first two issues.)
* Rails 5.1. stopped generating default DOM ids in the form helpers, and stopped putting the `new_class` HTML class on forms. There may be other differences that I've forgotten already
* Rails 5.2 returns to putting default DOM ids, but still doesn't put the HTML class on forms
* Given this history, it's reasonable to expect that future versions of Rails will make other changes that would affect `bootstrap_form`

Some of the guidelines I'm using for testing are:

* Tests should run reasonably quickly. I have a fast laptop but it's already almost two years old, and it runs all the tests in just over a second, so I think we're a ways from having a test suite that runs too long
* Tests should cover all variations in Rails versions that `bootstrap_form` currently supports
* Tests should cover all the variations of input to the gem. For a gem this big, that's of course impossible, but covering more will always be better than covering less
* Do the above in a way that's easy for current and future contributors. This means:
    * Try to concentrate the version handling in as few places as possible
    * Make it just as easy to use the helpers to test multiple versions, as it is to test one version

Reply:
Still do not understand the need to mess with `assert_equivalent_xml` by introducing helpers that change passed-in html. Tests themselves should send proper html that should be asserted. To illustrate:

```
# base test
assert expected_html, assert_equivalent_xml(form_for_output)

# good test
assert different_html, assert_equivalent_xml(form_with_output)

# what you're seemingly are doing
assert expected_html, assert_equivalent_xml(mangle_html(form_with_output))
```

At least that's what I think is happening.